adoc/extensions/sycl_khr_free_function_commands.adoc

```diff
@@ -0,0 +1,967 @@
+    size_t j = it.get_global_id(1);
+    size_t index = i * M + j;
+    c[index] = a[index] + b[index];
+  });
+
+  // Wait for all three kernels to complete before accessing the results.
+  // This blocks the host until all previous kernels have completed.
+  myQueue.wait();
+
+  std::cout << std::endl << "Result:" << std::endl;
+  for (size_t i = 0; i < N; i++) {
+    for (size_t j = 0; j < M; j++) {
+      size_t index = i * M + j;
+      // Compare the result to the analytic value.
+      if (c[index] != i * (2 + 2014) + j * (1 + 42)) {
+        std::cout << "Wrong value " << c[index] << " on element " << i << " "
+                  << j << std::endl;
+        exit(-1);
+      }
+    }
+  }
+
+  std::cout << "Good computation!" << std::endl;
+  return 0;
+}
+----
+
+== Requirements
+
+An instance of the [code]#requirements# class template encapsulates all of the
+requirements for scheduling a command.
+
+[source,role=synopsis]
+----
+namespace sycl::khr {
+
+template <typename T>
+struct is_requirement;
+
+template <typename T>
+static constexpr bool is_requirement_v = is_requirement<T>::value_type;
+
+template <typename... Requirements>
+class requirements {
+public:
+  // Only available if: all Requirements satisfy is_requirement_v
+  requirements(Requirements... values);
+
+  void register-accessors(handler& h) const;     // exposition only
+  void register-events(handler& h) const;        // exposition only
+  void register-kernel-bundle(handler& h) const; // exposition only
+  bool has-tracking() const;                     // exposition only
+};
+
+class tracking {
+public:
+  tracking(bool enabled=true);
+};
+
+}
+----
+
+Each instance of a type listed below defines a specific scheduling requirement.
+For each type, the [code]#is_requirement# type trait is specialized such that
+[code]#is_requirement_v<T># returns [code]#true#.
+
+* [code]#event#: The command must not begin executing until the event is
+  complete.
+
+* [code]#std::vector<event>#: The command must not begin executing until all
+  events in the vector are complete.
+
+* [code]#accessor#: The command must not begin executing until the
+  [code]#buffer# associated with the [code]#accessor# can be accessed in a
+  manner compatible with the specified [code]#access_mode#.
+
+* [code]#tracking#: The command must be submitted such that its state can be
+  tracked via an [code]#event# when the [code]#tracking# object is constructed
+  with an [code]#enabled# value of [code]#true#.
+
+* [code]#kernel_bundle<bundle_state::executable>#: The command must be submitted
+  using a <<device image>> from the kernel bundle.
+
+'''
+
+.[apititle]#Default constructor#
+[source,role=synopsis,id=api:requirements-ctor]
+----
+template <typename... Requirements>
+requirements(Requirements... values);
+----
+
+_Constraints_:
+* [code]#is_requirement_v# returns [code]#true# for each type in
+  [code]#Requirements#; and
+* [code]#Requirements# contains at most one
```

Comment:
Should we also have a constraint that there is at most one `tracking`, as these could have conflicting values?

Reply:
Do we know how to implement the constraint to forbid multiple occurrences of the same requirement object?  In order for the constraint to work, I think the type system needs to somehow count the number of `tracking` objects.  How would we do that?

Reply:
I'm pretty sure you can do this with some simple metaprogramming.

I can think of a few different ways that would work.  An implementation could filter the parameter pack to make multiple tuples (e.g., one of all the `accessor` objects, one of all the `event` objects), and then just `static_assert` if there was more than one `tracking` object.  Or an implementation could use a dedicated fold expression to increment some counter for each `tracking` object that it sees, and `static_assert` if that counter is greater than 1.

Reply:
Would it be easier to forbid duplicates when the `requirements` object is constructed?  (This is different from your current wording, where it is legal to construct a `requirements` with duplicates, but then place a constraint on the functions that take it.)  We should consider the complexity of the metaprogramming that is required in each case because that will affect compile time.

Reply:
These constraints are on the `requirements` object constructor, so the intent was to do exactly as you say.

We could collect some data here, but I don't think the complexity of the metaprogramming is very high, and I would hope that the overhead is minimal. The number of requirements is likely to be very low in the common case (e.g., a handful of accessors, tracking turned on, or a few dependent events).

Implementations already do this sort of thing and I haven't heard any complaints -- for example, I know that DPC++ uses a lot of metaprogramming to separate out different kinds of `reduction` objects.

Reply:
Oh, sorry.  I was confusing this constraint with the constraints below on the new free functions.